pyvo/extensions/cadcauth.py

```diff
@@ -0,0 +1,29 @@
+import logging
+import os
+
+import requests
+import requests.auth
+
+class CADCAuth(requests.auth.AuthBase):
+
+    def __init__(self):
+        self.cookie = None
+
+    def login(self, user, password, url='https://ws-cadc.canfar.net/ac/login'):
+        headers = {
+            'Content-Type': 'application/x-www-form-urlencoded',
+            'Accept': 'text/plain'
+        }
+
+        data = {
+            'username': user,
+            'password': password
+        }
+
+        response = requests.post(url, data=data, headers=headers)
+        response.raise_for_status()
+        self.cookie = '\"' + response.text + '\"'
+
+    def __call__(self, request):
+        request.prepare_cookies({'CADC_SSO': self.cookie})
```

Comment:
Technically, when you request a cookie, that cookie is valid only for that domain. That’s for user tracking prevention but in this case the cookie can be used to authenticate the user so we shouldn’t send it to just any site. Maybe add a valid domain list? At the CADC we made the browser request the cookie from all our domains in order to be valid across all of them.

Reply:
That's how the Gaia one works, but that's because it does the Set-Cookie headers.  For the CADC one, at least at this endpoint, it doesn't return the domains.  I'm happy to do this better, if I can find a better way of doing it.

This is somewhat related to the point I was making in my email about how we have two ways to determine when to give up the cookie: there's the capabilities that say cookie, and the RFC cookie with a domain.  Technically, in this way, once you have the cookie, you should always present it for that domain, even if the capabilities don't say to.  That's where this confusion comes in for me.

It's worth it to point out that it only uses this cookies for sessions attached to this authsession - it won't share them across anything else.  In this way, you have to be pretty explicit about which creds to use with which service.  For other auth methods, like client certificate or basic auth, there's nothing in the auth method that tells you when to use it or not, unlike cookies.  So I think in the end, we have to make a set of creds and associate them with exactly what we hope to use them with.  Even then, it won't present those creds unless the capabilities say to.

Reply:
(Sorry just in case I was obtuse, in the Gaia one, since the Set-Cookie headers are used, requests automatically knows to put them in the cookiejar.  request.cookies in that context means the jar, which requests will look at the domains and parse them out and only present them for the right things, as you are talking about)